lib/puppet/type/kdbkey.rb

```diff
@@ -0,0 +1,398 @@
+# encoding: UTF-8
+##
+# @file
+#
+# @brief Custom puppet type kdbkey for managing libelektra keys
+#
+# @copyright BSD License (see LICENSE or http://www.libelektra.org)
+#
+#
+require 'puppet/parameter/boolean'
+
+Puppet::Type.newtype(:kdbkey) do
+  @doc = <<-EOT
+    Manage libelekra keys.
+
+    This resource type allows to define and manipulate keys of libelektra's
+    key database.
+    EOT
+
+  feature :user, "ability to define/modify keys in the context of a specific user"
+
+
+  ensurable
+
+  # prefix parameter
+  #
+  # Note: this has to be defined BEFORE the name parameter, since we reference
+  # this prefix parameter within the names 'muge' and 'validate' methods
+  newparam(:prefix) do
+    desc <<-EOT
+    Prefix for the key name (optional)
+
+    If given, this value will prefix the given libelektra key name.
+    e.g.:
+
+      kdbkey { 'puppet/x1':
+        prefix => 'system/test',
+        value  => 'hello'
+      }
+
+    This will manage the key 'system/test/puppet/x1'.
+
+    Prefix and name are joined with a '/', if prefix does not end with '/'
+    or name does not start with '/'.
+
+    Both, name and prefix parameter are used to uniquely identify a
+    libelektra key.
+    EOT
+
+    isnamevar
+
+    defaultto ""
+
+    validate do |name|
+      unless name.nil? or name.empty?
+        unless name =~ /^(\/|spec|proc|dir|user|system)/
+          raise ArgumentError, "'%s' is not a valid basename" % name
```

Comment:
error is confusing, why basename? do you mean namespace?

It would be better if you get the list of namespaces from the binding (and in the binding from a loop using keyGetNamespace)